pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotTaskManager.java

```diff
@@ -47,115 +67,422 @@
+        if (!existingScheduledTasks.containsKey(newTaskType)) {
+          String newCronExpression = entry.getValue();
+          try {
+            scheduleJob(tableWithType, newTaskType, newCronExpression);
+          } catch (SchedulerException e) {
+            LOGGER.error("Failed to schedule cron task for table {}, task {}, cron expr {}", tableWithType, newTaskType,
+                newCronExpression, e);
+          }
+        }
+      }
+    } else {
+      for (String taskType : taskToCronExpressionMap.keySet()) {
+        // Schedule new job
+        String cronExpr = taskToCronExpressionMap.get(taskType);
+        try {
+          scheduleJob(tableWithType, taskType, cronExpr);
+        } catch (SchedulerException e) {
+          LOGGER.error("Failed to schedule cron task for table {}, task {}, cron expr {}", tableWithType, taskType,
+              cronExpr, e);
+        }
+      }
+    }
+    _tableTaskTypeToCronExpressionMap.put(tableWithType, taskToCronExpressionMap);
+  }
+
+  private void scheduleJob(String tableWithType, String taskType, String cronExprStr)
+      throws SchedulerException {
+    boolean exists = false;
+    try {
+      exists = _scheduler.checkExists(JobKey.jobKey(tableWithType, taskType));
+    } catch (SchedulerException e) {
+      LOGGER.error("Failed to check job existence for job key - table: {}, task: {} ", tableWithType, taskType, e);
+    }
+    if (!exists) {
+      LOGGER.info("Trying to schedule a job with cron expression: {} for table {}, task type: {}", cronExprStr,
+          tableWithType, taskType);
+      Trigger trigger = TriggerBuilder.newTrigger().withIdentity(TriggerKey.triggerKey(tableWithType, taskType))
+          .withSchedule(CronScheduleBuilder.cronSchedule(cronExprStr)).build();
+      JobDataMap jobDataMap = new JobDataMap();
+      jobDataMap.put(PINOT_TASK_MANAGER_KEY, this);
+      jobDataMap.put(LEAD_CONTROLLER_MANAGER_KEY, _leadControllerManager);
+      JobDetail jobDetail = JobBuilder.newJob(CronJobScheduleJob.class).withIdentity(tableWithType, taskType)
+          .setJobData(jobDataMap).build();
+      try {
+        _scheduler.scheduleJob(jobDetail, trigger);
+        _controllerMetrics.addValueToTableGauge(getCronJobName(tableWithType, taskType),
+            ControllerGauge.CRON_SCHEDULER_JOB_SCHEDULED, 1L);
+      } catch (Exception e) {
+        LOGGER.error("Failed to parse Cron expression - " + cronExprStr, e);
+        throw e;
+      }
+      Date nextRuntime = trigger.getNextFireTime();
+      LOGGER.info("Scheduled task for table: {}, task type: {}, next runtime: {}", tableWithType, taskType,
+          nextRuntime);
+    }
+  }
+
+  private Map<String, String> getTaskToCronExpressionMap(Map<String, Map<String, String>> taskTypeConfigsMap) {
+    Map<String, String> taskToCronExpressionMap = new HashMap<>();
+    for (String taskType : taskTypeConfigsMap.keySet()) {
+      Map<String, String> taskTypeConfig = taskTypeConfigsMap.get(taskType);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.